app/src/main/java/com/perflyst/twire/chat/ChatManager.java

```diff
@@ -435,15 +435,23 @@ private void handleNotice(IRCMessage message) {
      * If the roomstate has changed since last check variables are changed and the chatfragment is notified
      */
     private void handleRoomstate(IRCMessage message) {
-        boolean newR9k = message.tags.get("r9k").equals("1");
-        boolean newSlow = !message.tags.get("slow").equals("0");
-        boolean newSub = message.tags.get("subs-only").equals("1");
-        // If the one of the roomstate types have changed notify the chatfragment
-        if (chatIsR9kmode != newR9k || chatIsSlowmode != newSlow || chatIsSubsonlymode != newSub) {
-            chatIsR9kmode = newR9k;
-            chatIsSlowmode = newSlow;
-            chatIsSubsonlymode = newSub;
+        boolean roomstateChanged = false;
+
+        if( message.tags.get("r9k") != null) {
+            chatIsR9kmode = message.tags.get("r9k").equals("1");
+            roomstateChanged = true;
+        }
+        if( message.tags.get("slow") != null) {
+            chatIsSlowmode = !message.tags.get("slow").equals("0");
+            roomstateChanged = true;
+        }
+        if( message.tags.get("subs-only") != null) {
+            chatIsSubsonlymode = message.tags.get("subs-only").equals("1");
+            roomstateChanged = true;
+        }
 
+        // If the one of the roomstate types have changed notify the chatfragment
+        if (roomstateChanged) {
             onProgressUpdate(new ProgressUpdate(ProgressUpdate.UpdateType.ON_ROOMSTATE_CHANGE));
         }
```

Comment:
After looking at the [documentation](https://dev.twitch.tv/docs/irc/tags#roomstate-twitch-tags), I realized that the roomstate is only sent on joining or with a change. So we don't need to check to see if the roomstate has changed.
```suggestion
        if (message.tags.get("r9k") != null) {
            chatIsR9kmode = message.tags.get("r9k").equals("1");
        }
        if (message.tags.get("slow") != null) {
            chatIsSlowmode = !message.tags.get("slow").equals("0");
        }
        if (message.tags.get("subs-only") != null) {
            chatIsSubsonlymode = message.tags.get("subs-only").equals("1");
        }

        onProgressUpdate(new ProgressUpdate(ProgressUpdate.UpdateType.ON_ROOMSTATE_CHANGE));
```

Reply:
I think the check is still necessary because the roomstate still can be changed to one that won't affect the icons in the statusbar. Those states would be `emote-only` and `followers-only`. Changing in and out of emote-only-mode can be quite common, as this was the reason why I noticed the crash in the first place. Unnecessarily firing the `ON_ROOMSTATE_CHANGE` event certainly won't break anything but it isn't pretty either.